File: src/Server/base/Procedimiento.js
const express = require('express');
const db = require('./connection');
const router = express.Router();
const Joi = require("joi");
const util = require("util");

// Asincronizar consultas de la base de datos
const queryAsync = util.promisify(db.query).bind(db);

// Esquema de validación con Joi
const recetaInstruccionSchema = Joi.object({
  id_receta: Joi.number().integer().required(),
  instruccion: Joi.string().trim().min(3).required(),
  orden: Joi.number().integer().min(1).required(),
  Id_Usuario_Alta: Joi.number().integer().required(),
  Fecha_Alta: Joi.date().iso().required()
});

// Esquema para actualizar
const updateSchema = Joi.object({
  instruccion: Joi.string().trim().min(3).required(),
  orden: Joi.number().integer().min(1).required(),
  Id_Usuario_Alta: Joi.number().integer().required(),
  Fecha_Alta: Joi.date().iso().required()
});

// Crear una nueva instrucción de una receta
router.post("/", (req, res) => {
  const id_receta = 1; // Temporal, reemplazar con el real

  // Validar con Joi
  const { error, value } = recetaInstruccionSchema.validate({ ...req.body, id_receta });
  if (error) return res.status(400).json({ error: error.details[0].message });

  const { instruccion, orden, Id_Usuario_Alta, Fecha_Alta } = value;

  const query = `
    INSERT INTO receta_instrucciones (Id_Receta, Instruccion, Orden, Id_Usuario_Alta, Fecha_Alta) 
    VALUES (?, ?, ?, ?, ?)
  `;
  db.query(query, [id_receta, instruccion, orden, Id_Usuario_Alta, Fecha_Alta], (err, result) => {
    if (err) {
      console.error("Error al insertar instrucción:", err);
      return res.status(500).send("Error al agregar instrucción");
    }
    res.json({ id: result.insertId, message: "Instrucción agregada con éxito" });
  });
});

// Obtener todas las instrucciones de una receta específica
router.get("/:id/:id_usu", async (req, res) => {
  const { id, id_usu } = req.params;

  try {
    // Obtener autor de la receta
    const recetaAutorRes = await queryAsync(
      `SELECT Id_Usuario_Alta FROM receta WHERE Id_Receta = ?`,
      [id]
    );

    if (recetaAutorRes.length === 0) {
      return res.status(404).json({ error: "Receta no encontrada" });
    }

    const idAutor = recetaAutorRes[0].Id_Usuario_Alta;

    // Obtener rol del usuario que intenta acceder
    const usuarioRes = await queryAsync(
      `SELECT Id_Rol FROM usuario WHERE Id_Usuario = ?`,
      [id_usu]
    );

    if (usuarioRes.length === 0) {
      return res.status(404).json({ error: "Usuario no encontrado" });
    }

    const rol = usuarioRes[0].Id_Rol;

    // Verificar permisos:
    const esAutor = parseInt(id_usu) === idAutor;
    const esAdmin = rol === 2;
    const recetaEsPublica = idAutor === 1;

    if (!esAutor && !esAdmin && !recetaEsPublica) {
      console.log("Acceso denegado, sin permisos");
      return res.status(403).json({ error: "Acceso denegado, sin permisos" });
    }

    // Obtener datos de la receta
    const query1 = `
      SELECT 
      ri.Instruccion AS Nombre,
      ri.Orden AS Orden,
      ri.Id_Receta_Instrucciones AS id,
      ri.Activo AS Activo
    FROM receta_instrucciones ri
    WHERE ri.Id_Receta = ?
    ORDER BY ri.Orden ASC;
    `;

    const recetaInfo = await queryAsync(query1, [id]);

    if (recetaInfo.length === 0) {
      return res.status(404).json({ error: "Procedimiento no encontrado" });
    }

    res.json(recetaInfo);

  } catch (err) {
    console.error("Error al obtener procedimiento:", err);
    res.status(500).send("Error al obtener procedimiento");
  }
});


// Actualizar una instrucción de receta
router.put("/:id", (req, res) => {
  const { id } = req.params;
  if (!/^\d+$/.test(id)) return res.status(400).send("ID inválido");

  const { error, value } = updateSchema.validate(req.body);
  if (error) return res.status(400).json({ error: error.details[0].message });

  const { instruccion, orden, Id_Usuario_Alta, Fecha_Alta } = value;

  const query = `
    UPDATE receta_instrucciones
    SET Instruccion = ?, Orden = ?, Id_Usuario_Alta = ?, Fecha_Alta = ?
    WHERE Id_Receta_Instrucciones = ?
  `;
  db.query(query, [instruccion, orden, Id_Usuario_Alta, Fecha_Alta, id], (err, result) => {
    if (err) {
      console.error("Error al actualizar la instrucción:", err);
      return res.status(500).send("Error al actualizar la instrucción");
    }
    if (result.affectedRows === 0) {
      return res.status(404).send("Instrucción no encontrada");
    }
    res.json({ message: "Instrucción actualizada con éxito" });
  });
});

// Eliminar una instrucción de receta
router.delete("/:id", (req, res) => {
  const { id } = req.params;
  if (!/^\d+$/.test(id)) return res.status(400).send("ID inválido");

  const query = "DELETE FROM receta_instrucciones WHERE Id_Receta_Instrucciones = ?";
  
  db.query(query, [id], (err, result) => {
    if (err) {
      console.error("Error al eliminar la instrucción:", err);
      return res.status(500).send("Error al eliminar la instrucción");
    }
    if (result.affectedRows === 0) {
      return res.status(404).send("Instrucción no encontrada");
    }
    res.json({ message: "Instrucción eliminada con éxito" });
  });
});

module.exports = router;

File: src/serverP/base/Procedimiento.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from datetime import datetime
from .connection import get_db_connection
from typing import List

procedimiento_bp = APIRouter()

class InstruccionSchema(BaseModel):
    instruccion: str
    orden: int
    id_usuario_alta: int
    fecha_alta: datetime

# Crear una nueva instrucción
@procedimiento_bp.post("/")
async def crear_instruccion(instruccion: InstruccionSchema):
    id_receta = 1  # ID temporal hasta obtener el real
    query = """
        INSERT INTO receta_instrucciones (Id_Receta, Instruccion, Orden, Id_Usuario_Alta, Fecha_Alta)
        VALUES (?, ?, ?, ?, ?)
    """
    values = (id_receta, instruccion.instruccion, instruccion.orden, instruccion.id_usuario_alta, instruccion.fecha_alta)
    
    conn = get_db_connection()
    cursor = conn.cursor()
    try:
        cursor.execute(query, values)
        conn.commit()
        return {"id": cursor.lastrowid, "message": "Instrucción agregada con éxito"}
    except Exception as e:
        conn.rollback()
        raise HTTPException(status_code=500, detail=f"Error al agregar instrucción: {str(e)}")
    finally:
        cursor.close()
        conn.close()

# Obtener todas las instrucciones de una receta específica
@procedimiento_bp.get("/{id}", response_model=List[InstruccionSchema])
async def obtener_instrucciones(id: int):
    query = """
        SELECT Instruccion AS instruccion, Orden AS orden, Id_Usuario_Alta AS id_usuario_alta, Fecha_Alta AS fecha_alta
        FROM receta_instrucciones
        WHERE Id_Receta = ?
        ORDER BY Orden ASC
    """
    
    conn = get_db_connection()
    cursor = conn.cursor()
    try:
        cursor.execute(query, (id,))
        results = cursor.fetchall()
        return [InstruccionSchema(**dict(zip([column[0] for column in cursor.description], row))) for row in results]
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error al obtener instrucciones: {str(e)}")
    finally:
        cursor.close()
        conn.close()

# Actualizar una instrucción
@procedimiento_bp.put("/{id}")
async def actualizar_instruccion(id: int, instruccion: InstruccionSchema):
    query = """
        UPDATE receta_instrucciones
        SET Instruccion = ?, Orden = ?, Id_Usuario_Alta = ?, Fecha_Alta = ?
        WHERE Id_Receta_Instrucciones = ?
    """
    values = (instruccion.instruccion, instruccion.orden, instruccion.id_usuario_alta, instruccion.fecha_alta, id)
    
    conn = get_db_connection()
    cursor = conn.cursor()
    try:
        cursor.execute(query, values)
        conn.commit()
        if cursor.rowcount == 0:
            raise HTTPException(status_code=404, detail="Instrucción no encontrada")
        return {"message": "Instrucción actualizada con éxito"}
    except Exception as e:
        conn.rollback()
        raise HTTPException(status_code=500, detail=f"Error al actualizar instrucción: {str(e)}")
    finally:
        cursor.close()
        conn.close()

# Eliminar una instrucción
@procedimiento_bp.delete("/{id}")
async def eliminar_instruccion(id: int):
    query = "DELETE FROM receta_instrucciones WHERE Id_Receta_Instrucciones = ?"
    
    conn = get_db_connection()
    cursor = conn.cursor()
    try:
        cursor.execute(query, (id,))
        conn.commit()
        if cursor.rowcount == 0:
            raise HTTPException(status_code=404, detail="Instrucción no encontrada")
        return {"message": "Instrucción eliminada con éxito"}
    except Exception as e:
        conn.rollback()
        raise HTTPException(status_code=500, detail=f"Error al eliminar instrucción: {str(e)}")
    finally:
        cursor.close()
        conn.close()
